python/cudf_polars/cudf_polars/dsl/expressions/rolling.py

```diff
@@ -210,10 +216,317 @@ def __init__(
+        ).columns()[0]
+        max_pos = plc.copying.gather(
+            plc.Table([pos_in_group]),
+            run_ends,
+            plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+        ).columns()[0]
+        min_pos = plc.binaryop.binary_operation(
+            min_pos, one, plc.binaryop.BinaryOperator.ADD, size_type
+        )
+        max_pos = plc.binaryop.binary_operation(
+            max_pos, one, plc.binaryop.BinaryOperator.ADD, size_type
+        )
+
+        def _per_row_ranks_from_run_pos(positions: plc.Column) -> plc.Column:
+            # Map the per-run min/max ranks back to per-row results
+            run_ids_at_starts = plc.copying.gather(
+                plc.Table([run_id_sorted]),
+                run_starts,
+                plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+            ).columns()[0]
+            run_index_per_row = plc.binaryop.binary_operation(
+                run_id_sorted, one, plc.binaryop.BinaryOperator.SUB, size_type
+            )
+            run_index_per_run = plc.binaryop.binary_operation(
+                run_ids_at_starts, one, plc.binaryop.BinaryOperator.SUB, size_type
+            )
+
+            mapping = plc.copying.scatter(
+                plc.Table([positions]),
+                run_index_per_run,
+                plc.Table(
+                    [
+                        plc.Column.from_scalar(
+                            plc.Scalar.from_py(0, size_type), num_tied_runs
+                        )
+                    ]
+                ),
+            ).columns()[0]
+
+            ranks_sorted = plc.copying.gather(
+                plc.Table([mapping]),
+                run_index_per_row,
+                plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+            ).columns()[0]
+
+            return plc.copying.gather(
+                plc.Table([ranks_sorted]),
+                inv_perm_i_to_k,
+                plc.copying.OutOfBoundsPolicy.DONT_CHECK,
+            ).columns()[0]
+
+        if method == "min":
+            return GroupedRollingWindow._rebuild_col_with_nulls(
+                _per_row_ranks_from_run_pos(min_pos), i_seq, n, out_dtype
+            )
+        if method == "max":
+            return GroupedRollingWindow._rebuild_col_with_nulls(
+                _per_row_ranks_from_run_pos(max_pos), i_seq, n, out_dtype
+            )
+        if method == "average":
+            max_ranks = _per_row_ranks_from_run_pos(max_pos)
+            min_ranks = _per_row_ranks_from_run_pos(min_pos)
+            f64 = plc.DataType(plc.TypeId.FLOAT64)
+            min_ = plc.unary.cast(min_ranks, f64)
+            max_ = plc.unary.cast(max_ranks, f64)
+            sum_ = plc.binaryop.binary_operation(
+                min_, max_, plc.binaryop.BinaryOperator.ADD, f64
+            )
+            two = plc.Scalar.from_py(2.0, f64)
+            avg = plc.binaryop.binary_operation(
+                sum_, two, plc.binaryop.BinaryOperator.DIV, f64
+            )
```

Comment:
Subtle point:

From the polars docs,
> The average of the ranks that would have been assigned to all the tied values is assigned to each value.

Say we have `m` ties; starting with tie `k` then the average of the tied ranks is:

$= \frac{\sum_{i=0}^{m-1}{k=i}}{m}$ = $\frac{(k+(k+m-1))((k+m-1)-k+1)}{2m} = \frac{k + (k+m-1)}{2} = \frac{min+max}{2}$